flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java

```diff
@@ -125,6 +121,16 @@ public String getClusterDescription() {
         };
     }
 
+    private String getWebMonitorAddress(Configuration configuration) throws Exception {
+        HighAvailabilityServicesUtils.AddressResolution resolution =
+                HighAvailabilityServicesUtils.AddressResolution.TRY_ADDRESS_RESOLUTION;
+        if (configuration.get(KubernetesConfigOptions.REST_SERVICE_EXPOSED_TYPE)
+                == KubernetesConfigOptions.ServiceExposedType.ClusterIP) {
+            resolution = HighAvailabilityServicesUtils.AddressResolution.NO_ADDRESS_RESOLUTION;
+        }
+        return HighAvailabilityServicesUtils.getWebMonitorAddress(configuration, resolution);
+    }
```

Comment:
Instead of suppressing the web monitor address resolution, can't we tell the `RestClusterClient` the address to the namespaced service? I think you mentioned in the ticket that we will communicate with the cluster through this service if we have chose `ClusterIP`. It just feels wrong that we still retrieve the web monitor's address even though we don't need it.

Reply:
Flink client communicates with the cluster via namespaced service if ClusterIP is chosen.

I assume you mean directly return a `RestClusterClient` using the namespaced service(aka `restEndpoint.get().getAddress()`). After then, we also need to check whether the ssl is enabled and add `http/https` protocol. I think it is what we have done in `HighAvailabilityServicesUtils.getWebMonitorAddress`.

Moreover, I do not think we are retrieving the web monitor's address. It is more like to construct the address in a specific schema(aka protocol://address:port). The retrieval process has already been done in the `flinkKubeClient.getRestEndpoint`.

Reply:
In the K8s case, from where exactly do we retrieve the address of the service? If I understood you correctly, then `RestOptions.ADDRESS` contains some address which is not resolvable from the outside. Hence, I am wondering why we should try to construct the web monitor address from this configuration at all.

Reply:
After a little more consideration, maybe we always do not need to resolve the rest endpoint address.
When the service exposed type is NodePort/LoadBalancer, the rest endpoint address is usually a ip address, not a hostname. For ClusterIP, we also do not need to resolve the address.

Reply:
The Flink application submission could happen in or out of the K8s cluster. The reason why we set the `RestOptions.ADDRESS` to the namespaced service is that it could be directly used in the K8s cluster. However, when the submission happens out of the K8s cluster, the namespaced service could not be used to contact with the cluster. 

In such situation, users usually need to create a ingress for the communication.

Reply:
Assuming the submission happens outside of the K8s cluster with `ClusterIP` configured, does this mean that the `RestClusterClient` cannot talk to the cluster? If yes, then there is not really a need for creating it, does it? But then there is the problem that we could also deploy the cluster from within the K8s cluster where the `RestClusterClient` can talk to the Flink cluster, right?

Should we maybe say that user's have to submit the cluster in detached mode if they cannot connect to it? The problem with the current solution is that we might start an attached per job cluster with which we cannot talk. Hence, we cannot request the final job result either and consequently, the cluster will never shut down. So maybe we should fail with an better exception message, what do you think?